market-contract/src/nft_callbacks.rs

```diff
@@ -38,37 +38,21 @@ impl NonFungibleTokenApprovalsReceiver for Contract {
     ) {
         // get the contract ID which is the predecessor
         let nft_contract_id = env::predecessor_account_id();
-        //get the signer which is the person who initiated the transaction
-        let signer_id = env::signer_account_id();
-
-        //make sure that the signer isn't the predecessor. This is so that we're sure
-        //this was called via a cross-contract call
-        assert_ne!(
-            nft_contract_id,
-            signer_id,
-            "nft_on_approve should only be called via cross-contract call"
-        );
-        //make sure the owner ID is the signer. 
-        assert_eq!(
-            owner_id,
-            signer_id,
-            "owner_id should be signer_id"
-        );
 
         //we need to enforce that the user has enough storage for 1 EXTRA sale.  
 
         //get the storage for a sale. dot 0 converts from U128 to u128
         let storage_amount = self.storage_minimum_balance().0;
         //get the total storage paid by the owner
-        let owner_paid_storage = self.storage_deposits.get(&signer_id).unwrap_or(0);
+        let owner_paid_storage = self.storage_deposits.get(&nft_contract_id).unwrap_or(0);
```

Comment:
The storage isn't for the NFT contract and should belong to the person or organization that is listing the NFT. By removing the signer, I'm not too sure who should be responsible for this payment. Perhaps it would be the signer still or maybe the token owner.

Reply:
It cannot be the owner, since it's a parameter, otherwise a third party could simply drain your storage by setting you as the owner.

Since the one adding storage is the predecessor, the predecessor needs to pay to hamper possible attacks.

Reply:
Maybe we need to transform this into payable and make the marketplace deposit money, which is in turn asked from the user.

Reply:
We can't do this as it's not part of the standard. I think we just keep the signer and use it in this one case but remove the assertion.

Reply:
There are two possible solutions:
1. We make `nft_on_approve` to callback the predecessor and make sure the owner is effectively the NFT owner
2. We remove `nft_on_approve` and transform it into "put_nft_on_sale", which checks if the predecessor is the owner of the NFT.

I personally prefer the 2nd one.

Reply:
I don't want to overly complicate this contract as again, it's meant to be a tutorial on how things could work on NEAR. For this reason I would like to keep the signer's storage check personally.